File: python/src/rql/explain.py
"""Human-readable / structured explanation of a PhysicalPlan."""
from __future__ import annotations

from typing import Any, Literal, Union


def explain(
    physical: dict[str, Any],
    *,
    format: Literal["text", "object"] = "text",
) -> Union[str, dict[str, Any]]:
    """Explain a PhysicalPlan as text or a structured object.

    Does not execute against any database.
    """
    if physical.get("kind") != "PhysicalPlan":
        raise ValueError(f"expected kind=PhysicalPlan, got {physical.get('kind')!r}")

    meta = physical.get("meta") or {}
    caps = physical.get("capabilitiesUsed") or {}
    budgets = physical.get("budgets") or {}
    root = physical.get("root") or {}

    lines: list[str] = []
    ops: list[dict[str, Any]] = []

    def walk(node: dict[str, Any], depth: int = 0) -> None:
        op = node.get("op", "?")
        nid = node.get("id", "")
        detail_parts: list[str] = []
        if op == "AnnExec":
            detail_parts.append(f"k={node.get('k')} index={node.get('index')} metric={node.get('metric')}")
        elif op == "Bm25Exec":
            detail_parts.append(f"k={node.get('k')}")
        elif op == "LateInteractExec":
            detail_parts.append(f"variant={node.get('variant')} k={node.get('k')}")
        elif op == "FilterExec":
            detail_parts.append(f"mode={node.get('mode')} pruning={node.get('pruningStrategy')}")
            pred = node.get("predicate") or {}
            if pred.get("aclHard"):
                detail_parts.append("aclHard")
        elif op == "FusionExec":
            detail_parts.append(f"family={node.get('family')} native={node.get('native')}")
        elif op == "ShimCast":
            detail_parts.append(f"shim={node.get('shim')} expensive={node.get('expensive')}")
        detail = " ".join(detail_parts)
        indent = "  " * depth
        lines.append(f"{indent}{op}({nid}){' ' + detail if detail else ''}")
        ops.append({"id": nid, "op": op, "depth": depth, "detail": detail})
        if isinstance(node.get("input"), dict):
            walk(node["input"], depth + 1)
        for child in node.get("inputs") or []:
            if isinstance(child, dict):
                walk(child, depth + 1)

    walk(root)

    obj: dict[str, Any] = {
        "kind": "PhysicalPlanExplain",
        "schemaVersion": physical.get("schemaVersion"),
        "profileId": meta.get("profileId"),
        "label": meta.get("label"),
        "notes": meta.get("notes"),
        "capabilitiesUsed": caps,
        "budgets": budgets,
        "ops": ops,
        "tree": "\n".join(lines),
        "notExecuted": True,
    }

    if format == "object":
        return obj

    out = [
        f"PhysicalPlan explain (profile={meta.get('profileId') or '?'})",
        f"schemaVersion={physical.get('schemaVersion')} label={meta.get('label')}",
    ]
    if meta.get("notes"):
        out.append(f"notes: {meta['notes']}")
    if budgets:
        out.append(f"budgets: {budgets}")
    out.append("tree:")
    out.extend(lines)
    out.append("(sketch / notExecuted — emit is docs-shaped until live adapters)")
    return "\n".join(out) + "\n"

File: javascript/src/explain.ts
/** Human-readable / structured explanation of a PhysicalPlan. */

export type ExplainFormat = "text" | "object";

export type ExplainObject = {
  kind: "PhysicalPlanExplain";
  schemaVersion?: unknown;
  profileId?: unknown;
  label?: unknown;
  notes?: unknown;
  capabilitiesUsed: Record<string, unknown>;
  budgets: Record<string, unknown>;
  ops: { id: unknown; op: unknown; depth: number; detail: string }[];
  tree: string;
  notExecuted: true;
};

export function explain(
  physical: Record<string, unknown>,
  options: { format?: ExplainFormat } = {},
): string | ExplainObject {
  if (physical.kind !== "PhysicalPlan") {
    throw new Error(`expected kind=PhysicalPlan, got ${JSON.stringify(physical.kind)}`);
  }
  const format = options.format ?? "text";
  const meta = (physical.meta || {}) as Record<string, unknown>;
  const caps = (physical.capabilitiesUsed || {}) as Record<string, unknown>;
  const budgets = (physical.budgets || {}) as Record<string, unknown>;
  const root = (physical.root || {}) as Record<string, unknown>;
  const lines: string[] = [];
  const ops: ExplainObject["ops"] = [];

  const walk = (node: Record<string, unknown>, depth = 0): void => {
    const op = String(node.op ?? "?");
    const nid = node.id ?? "";
    const detailParts: string[] = [];
    if (op === "AnnExec") detailParts.push(`k=${node.k} index=${node.index} metric=${node.metric}`);
    else if (op === "Bm25Exec") detailParts.push(`k=${node.k}`);
    else if (op === "LateInteractExec") detailParts.push(`variant=${node.variant} k=${node.k}`);
    else if (op === "FilterExec") {
      detailParts.push(`mode=${node.mode} pruning=${node.pruningStrategy}`);
      const pred = (node.predicate || {}) as Record<string, unknown>;
      if (pred.aclHard) detailParts.push("aclHard");
    } else if (op === "FusionExec") detailParts.push(`family=${node.family} native=${node.native}`);
    else if (op === "ShimCast") detailParts.push(`shim=${node.shim} expensive=${node.expensive}`);
    const detail = detailParts.join(" ");
    lines.push(`${"  ".repeat(depth)}${op}(${nid})${detail ? " " + detail : ""}`);
    ops.push({ id: nid, op, depth, detail });
    if (node.input && typeof node.input === "object") walk(node.input as Record<string, unknown>, depth + 1);
    for (const child of (node.inputs as Record<string, unknown>[] | undefined) || []) {
      if (child && typeof child === "object") walk(child, depth + 1);
    }
  };
  walk(root);

  const obj: ExplainObject = {
    kind: "PhysicalPlanExplain",
    schemaVersion: physical.schemaVersion,
    profileId: meta.profileId,
    label: meta.label,
    notes: meta.notes,
    capabilitiesUsed: caps,
    budgets,
    ops,
    tree: lines.join("\n"),
    notExecuted: true,
  };
  if (format === "object") return obj;

  const out = [
    `PhysicalPlan explain (profile=${meta.profileId ?? "?"})`,
    `schemaVersion=${physical.schemaVersion} label=${meta.label}`,
  ];
  if (meta.notes) out.push(`notes: ${meta.notes}`);
  if (Object.keys(budgets).length) out.push(`budgets: ${JSON.stringify(budgets)}`);
  out.push("tree:");
  out.push(...lines);
  out.push("(sketch / notExecuted — emit is docs-shaped until live adapters)");
  return out.join("\n") + "\n";
}
